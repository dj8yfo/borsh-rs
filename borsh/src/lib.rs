#![cfg_attr(not(feature = "std"), no_std)]

/*!

# Crate features

### Ecosystem features

* **std** -
  When enabled, this will cause `borsh` to use the standard library. Currently,
  disabling this feature will result in building the crate in `no_std` environment.

### Default features

* **std** - enabled by default.

### Other features

* **rc** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [`Rc<T>`](std::rc::Rc)/[`Arc<T>`](std::sync::Arc) respectively.
  In `no_std` setting `Rc`/`Arc` are pulled from `alloc` crate.
* **hashbrown** -
  Pulls in [HashMap](std::collections::HashMap)/[HashSet](std::collections::HashSet) when no `std` is available.
  This feature is set to be mutually exclusive with **std** feature.
* **bytes** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [Bytes](bytes::Bytes) and [BytesMut](bytes::BytesMut).
* **bson** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [ObjectId](bson::oid::ObjectId).


### Config aliases

* **hash_collections** -
  This is a feature alias, set up in `build.rs` to be equivalent to (**std** OR **hashbrown**).
  This alias gates [schema](crate::schema) and [schema_helpers](crate::schema_helpers) modules, as
  [BorshSchema](crate::schema::BorshSchema) relies on [HashMap](std::collections::HashMap) existing.
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [HashMap](std::collections::HashMap)/[HashSet](std::collections::HashSet).


*/

#[cfg(not(feature = "std"))]
extern crate alloc;

#[cfg(feature = "schema")]
pub use borsh_derive::BorshSchema;
pub use borsh_derive::{BorshDeserialize, BorshSerialize};

pub mod de;

// See `hash_collections` alias definition in build.rs
#[cfg(feature = "schema")]
pub mod schema;
#[cfg(feature = "schema")]
pub mod schema_helpers;
pub mod ser;

pub use de::BorshDeserialize;
pub use de::{from_reader, from_slice};
#[cfg(feature = "schema")]
pub use schema::BorshSchema;
#[cfg(feature = "schema")]
pub use schema_helpers::{try_from_slice_with_schema, try_to_vec_with_schema};
pub use ser::helpers::{to_vec, to_writer};
pub use ser::BorshSerialize;

#[cfg(all(feature = "std", feature = "hashbrown"))]
compile_error!("feature \"std\" and feature \"hashbrown\" don't make sense at the same time");

/// A facade around all the types we need from the `std`, `core`, and `alloc`
/// crates. This avoids elaborate import wrangling having to happen in every
/// module.
#[doc(hidden)]
#[cfg(feature = "std")]
pub mod __maybestd {
    pub use std::{borrow, boxed, collections, format, io, string, vec};

    #[cfg(feature = "rc")]
    pub use std::{rc, sync};
}

#[cfg(not(feature = "std"))]
pub mod nostd_io;

#[doc(hidden)]
#[cfg(not(feature = "std"))]
pub mod __maybestd {
    pub use alloc::{borrow, boxed, format, string, vec};

    #[cfg(feature = "rc")]
    pub use alloc::{rc, sync};

    pub mod collections {
        pub use alloc::collections::{btree_map, BTreeMap, BTreeSet, LinkedList, VecDeque};
        #[cfg(feature = "hashbrown")]
        pub use hashbrown::*;
    }

    pub mod io {
        pub use super::super::nostd_io::*;
    }
}
