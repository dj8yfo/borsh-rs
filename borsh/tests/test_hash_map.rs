#![cfg_attr(not(feature = "std"), no_std)]
#![cfg(hash_collections)]

#[cfg(feature = "std")]
use core::hash::BuildHasher;

#[cfg(feature = "hashbrown")]
use hashbrown::{HashMap, HashSet};
#[cfg(feature = "std")]
use std::collections::{
    hash_map::{DefaultHasher, RandomState},
    HashMap, HashSet,
};
#[cfg(not(feature = "std"))]
extern crate alloc;
#[cfg(not(feature = "std"))]
use alloc::string::{String, ToString};

use borsh::{from_slice, BorshSerialize};

fn default_hashset() -> HashSet<String> {
    let mut set = HashSet::new();
    set.insert("foo".to_string());
    set.insert("many".to_string());
    set.insert("various".to_string());
    set.insert("different".to_string());
    set.insert("keys".to_string());
    set.insert("one".to_string());
    set
}

fn default_hashmap() -> HashMap<String, String> {
    let mut map = HashMap::new();
    map.insert("foo".to_string(), "bar".to_string());
    map.insert("one".to_string(), "two".to_string());
    map.insert("123mayn".to_string(), "two".to_string());
    map.insert("_more".to_string(), "two".to_string());
    map.insert("diefferent".to_string(), "two".to_string());
    map
}

#[test]
fn test_default_hashmap() {
    let map = default_hashmap();

    let data = map.try_to_vec().unwrap();
    let actual_map = from_slice::<HashMap<String, String>>(&data).unwrap();
    assert_eq!(map, actual_map);
}

#[test]
fn test_default_hashset() {
    let set = default_hashset();

    let data = set.try_to_vec().unwrap();
    let actual_set = from_slice::<HashSet<String>>(&data).unwrap();
    assert_eq!(set, actual_set);
}

#[cfg(feature = "de_strict_order")]
const ERROR_WRONG_ORDER_OF_KEYS: &str = "keys were not serialized in ascending order";

#[cfg(feature = "de_strict_order")]
#[test]
fn test_hashset_deser_err_wrong_order() {
    let mut vec_str = vec![];
    vec_str.push("various".to_string());
    vec_str.push("foo".to_string());
    vec_str.push("many".to_string());

    let data = vec_str.try_to_vec().unwrap();
    let result = from_slice::<HashSet<String>>(&data);
    assert!(result.is_err());

    assert_eq!(result.unwrap_err().to_string(), ERROR_WRONG_ORDER_OF_KEYS);
}

#[cfg(feature = "de_strict_order")]
#[test]
fn test_hashmap_deser_err_wrong_order() {
    let val = "val".to_string();
    let mut vec_str = vec![];
    vec_str.push(("various".to_string(), val.clone()));
    vec_str.push(("foo".to_string(), val.clone()));
    vec_str.push(("many".to_string(), val.clone()));

    let data = vec_str.try_to_vec().unwrap();
    let result = from_slice::<HashMap<String, String>>(&data);
    assert!(result.is_err());

    assert_eq!(result.unwrap_err().to_string(), ERROR_WRONG_ORDER_OF_KEYS);
}

#[derive(Default)]
#[cfg(feature = "std")]
struct NewHasher(RandomState);

#[cfg(feature = "std")]
impl BuildHasher for NewHasher {
    type Hasher = DefaultHasher;
    fn build_hasher(&self) -> DefaultHasher {
        self.0.build_hasher()
    }
}

#[test]
#[cfg(feature = "std")]
fn test_generic_hash_hashmap() {
    let mut map = HashMap::with_hasher(NewHasher::default());
    map.insert("foo".to_string(), "bar".to_string());
    map.insert("one".to_string(), "two".to_string());

    let data = map.try_to_vec().unwrap();
    let actual_map = from_slice::<HashMap<String, String, NewHasher>>(&data).unwrap();
    assert_eq!(map, actual_map);
}

#[test]
#[cfg(feature = "std")]
fn test_generic_hashset() {
    let mut set = HashSet::with_hasher(NewHasher::default());
    set.insert("foo".to_string());
    set.insert("many".to_string());
    set.insert("various".to_string());
    set.insert("different".to_string());
    set.insert("keys".to_string());
    set.insert("one".to_string());

    let data = set.try_to_vec().unwrap();
    let actual_set = from_slice::<HashSet<String, NewHasher>>(&data).unwrap();
    assert_eq!(set, actual_set);
}
