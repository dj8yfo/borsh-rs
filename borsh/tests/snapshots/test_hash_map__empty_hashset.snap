---
source: borsh/tests/test_hash_map.rs
expression: data
---
[
    0,
    0,
    0,
    0,
]
